### PR TITLE
feat: add no-implicit-declare-namespace-export (#1057)

### LINKED ISSUE #1057: Lint for Implicitly Exported Types in Declare Namespace
Hello, as https://github.com/denoland/deno/issues/14976 has shown, it can be hard to realize TypeScript's default behavior of exporting all types in a `declare namespace` block.  It might be beneficial to add an optional lint (something like `no_implicit_declare_namespace_export`) onto `type` and `interface` members that are implicitly exported.  This should be able to help stop further costly bugs from happening in Deno and elsewhere.

Example code:

```ts
declare namespace ResourceHandler {
  type FormatResource<T extends string> = `resource-${T}`;
}
```

Output:

```
Type 'FormatResource' is implicitly exported from the 'ResourceHandler' namespace. Try adding an 'export {};' to the top of the namespace to disable this behavior.
```

And the suggestion could be integrated into editors.  Thanks

### PR DESCRIPTION
**Pull request by Conaclos**
Fixes #1057

### CODE REVIEW

File: src/rules/no_implicit_declare_namespace_export.rs [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+  fn lint_program(&self, _context: &mut Context, _program: ProgramRef<'_>) {
+    unreachable!();
+  }
+
+  fn lint_program_with_ast_view(
+    &self,
+    context: &mut Context,
+    program: Program<'_>,
+  ) {
+    if context.file_name().ends_with(".d.ts") {
```
Comment (ah-yu) [inline src/rules/no_implicit_declare_namespace_export.rs:39]: This is different from what is written in the documentation. The document said this lint rule works for both `ts` and `d.ts` files.

Comment (Conaclos) [reply src/rules/no_implicit_declare_namespace_export.rs]: Thanks! It is now fixed.

**Comment by Conaclos** [comment]
What is the next step for approval?

**Comment by bartlomieju** [comment]
Hey @Conaclos sorry for slow review, I missed previous notification.

Implementation wise it looks great to me, given that it's not on by default I think we can land it as is.

**Comment by Conaclos** [comment]
@bartlomieju No problem :) Great to hear it :)

**Pull request merged**